### PR TITLE
Restructure of lightning logging docs

### LINKED ISSUE #811: F1 compute metric warning
## 🐛 Bug

Get a warning in F1 metric compute after following documentation: torchmetrics/utilities/prints.py:36: UserWarning: The ``compute`` method of metric F1 was called before the ``update`` method which may lead to errors, as metric states have not yet been updated. I checked docs and tutorials carefully and all looks correct to me - any suggestions would be most appreciated! Many thanks for help.

### To Reproduce

Error happens in the line `f1 = self.valid_f1.compute()` below 

#### Code sample

```py
class LSTMTaggerModel(pl.LightningModule):
    def __init__(
        self,
        num_classes,
        class_map,
        from_checkpoint=False,
        model_name='last.ckpt',
        learning_rate=3e-6,
        **kwargs,
    ):

        super().__init__()
        self.save_hyperparameters()
        self.learning_rate = learning_rate
        self.model = LSTMClassifier(num_classes=num_classes)
        self.class_map = class_map
        self.num_classes = num_classes
        self.valid_acc = torchmetrics.Accuracy()
        self.valid_f1 = torchmetrics.F1()


    def forward(self, *input, **kwargs):
        return self.model(*input, **kwargs)

    def training_step(self, batch, batch_idx):
        x, y_true = batch
        loss, _ = self(x, labels=y_true)
        self.log('train_loss', loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y_true = batch
        _, y_pred = self(x, labels=y_true)
        preds = torch.argmax(y_pred, axis=1)
        self.valid_acc(preds, y_true[0])
        self.log('val_acc', self.valid_acc, prog_bar=True)
        self.valid_f1(preds, y_true[0])
        self.log('f1', self.valid_f1, prog_bar=True)     

    def configure_optimizers(self):
        'Prepare optimizer and schedule (linear warmup and decay)'
        opt = torch.optim.Adam(params=self.parameters(), lr=self.learning_rate)
        sch = torch.optim.lr_scheduler.CosineAnnealingLR(opt, T_max=10)
        return [opt], [sch]

    def training_epoch_end(self, training_step_outputs):
        avg_loss = torch.tensor([x['loss']
                                 for x in training_step_outputs]).mean()
        self.log('train_loss', avg_loss)
        print(f'###score: train_loss### {avg_loss}')

    def validation_epoch_end(self, val_step_outputs):
        acc = self.valid_acc.compute()
        f1 = self.valid_f1.compute()
        self.log('val_score', acc)
        self.log('f1', f1)
        print(f'###score: val_score### {acc}')

```

### Expected behavior

No warning is provided 

### Environment

- PyTorch Version (e.g., 1.0):
torchmetrics = "0.3.2"
pytorch-lightning = "1.2.7"
torch = "1.7.1"
- OS (e.g., Linux): MacOS
- How you installed PyTorch: with poetry 
- Python version: 3.7
- CUDA/cuDNN version: NA (running on cpu)
- GPU models and configuration: Apple Silicon M1

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #811
Restructuring of lightning logging docs. Currently consist of a bunch of notes which I tried to format into a more coherent text.
Added a "pitfalls" section with description of some common pitfalls.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: docs/source/pages/lightning.rst [hunk 1]
```diff
@@ -9,17 +9,21 @@
...
+.. note::
 
     TorchMetrics always offers compatibility with the last 2 major PyTorch Lightning versions, but we recommend to always keep both frameworks
     up-to-date for the best experience.
 
 While TorchMetrics was built to be used with native PyTorch, using TorchMetrics with Lightning offers additional benefits:
 
-* Module metrics are automatically placed on the correct device when properly defined inside a LightningModule. This means that your data will always be placed on the same device as your metrics.
-* Native support for logging metrics in Lightning using `self.log <https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule>`_ inside your LightningModule.
+* Module metrics are automatically placed on the correct device when properly defined inside a LightningModule.
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:22]:
```suggestion
* Modular metrics are automatically placed on the correct device when properly defined inside a LightningModule.
```

File: docs/source/pages/lightning.rst [hunk 2]
```diff
@@ -45,45 +49,19 @@ The example below shows how to use a metric in your `LightningModule <https://py
...
-                return [DataLoader(...), DataLoader(...)]
-
-            def validation_step(self, batch, batch_idx, dataloader_idx):
-                x, y = batch
-                preds = self(x)
-                ...
-                self.val_acc[dataloader_idx](preds, y)
-                self.log('val_acc', self.val_acc[dataloader_idx])
-
+Metric logging in Lightning happens through the `self.log` or `self.log_dict` method. Both methods only supports logging of *scalar-tensors*.
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:52]:
```suggestion
Metric logging in Lightning happens through the ``self.log`` or ``self.log_dict`` method. Both methods only support the logging of *scalar-tensors*.
```

File: docs/source/pages/lightning.rst [hunk 3]
```diff
@@ -9,17 +9,21 @@
...
 
     TorchMetrics always offers compatibility with the last 2 major PyTorch Lightning versions, but we recommend to always keep both frameworks
     up-to-date for the best experience.
 
 While TorchMetrics was built to be used with native PyTorch, using TorchMetrics with Lightning offers additional benefits:
 
-* Module metrics are automatically placed on the correct device when properly defined inside a LightningModule. This means that your data will always be placed on the same device as your metrics.
-* Native support for logging metrics in Lightning using `self.log <https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule>`_ inside your LightningModule.
+* Module metrics are automatically placed on the correct device when properly defined inside a LightningModule.
+  This means that your data will always be placed on the same device as your metrics. No need to call `.to(device)` anymore!
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:23]:
```suggestion
  This means that your data will always be placed on the same device as your metrics. No need to call ``.to(device)`` anymore!
```

File: docs/source/pages/lightning.rst [hunk 4]
```diff
@@ -45,45 +49,19 @@ The example below shows how to use a metric in your `LightningModule <https://py
...
 ********************
 Logging TorchMetrics
 ********************
 
-When :class:`~torchmetrics.Metric` objects, which return a scalar tensor,  are logged directly in Lightning using the LightningModule `self.log <https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule>`_ method. Lightning will log
-the metric based on ``on_step`` and ``on_epoch`` flags present in ``self.log(...)``.
-If ``on_epoch`` is True, the logger automatically logs the end of epoch metric value by calling
-``.compute()``.
+Logging metrics can be done in two ways: either logging the metric object directly or the computed metric values. When :class:`~torchmetrics.Metric` objects, which return a scalar tensor,
+are logged directly in Lightning using the LightningModule `self.log <https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule>`_ method.
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:62]:
```suggestion
Logging metrics can be done in two ways: either logging the metric object directly or the computed metric values. When :class:`~torchmetrics.Metric` objects, which return a scalar tensor
are logged directly in Lightning using the LightningModule `self.log <https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule>`_ method,
```

File: docs/source/pages/lightning.rst [hunk 5]
```diff
@@ -118,85 +96,125 @@ If ``on_epoch`` is True, the logger automatically logs the end of epoch metric v
...
+            ...
+            self.valid_acc.update(logits, y)
+
+        def validation_epoch_end(self, outputs):
+            self.log('valid_acc_epoch', self.valid_acc.compute())
+            self.valid_acc.reset()
+
+Note that logging metrics this ways will require you to manually reset the metrics at the end of the epoch yourself. In general we recommend logging
+the metric object to make sure that metrics are correctly computed and reset. Additionally, we highly recommend that the two ways of logging is not
+mixed as it can lead to wrong results.
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:132]:
```suggestion
Note that logging metrics this way will require you to manually reset the metrics at the end of the epoch yourself. In general, we recommend logging
the metric object to make sure that metrics are correctly computed and reset. Additionally, we highly recommend that the two ways of logging are not
mixed as it can lead to wrong results.
```

File: docs/source/pages/lightning.rst [hunk 6]
```diff
@@ -118,85 +96,125 @@ If ``on_epoch`` is True, the logger automatically logs the end of epoch metric v
...
 
-                # this will reset the metric automatically at the epoch end
-                self.train_acc(preds, y)
-                self.log('train_acc', self.train_acc, on_step=True, on_epoch=False)
 
-                # this will not reset the metric automatically at the epoch end
-                precision = self.train_precision(preds, y)
-                self.log('train_precision', precision, on_step=True, on_epoch=False)
+***************
+Common pitfalls
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:156]:
```suggestion
Common Pitfalls
```

File: docs/source/pages/lightning.rst [hunk 7]
```diff
@@ -45,45 +49,19 @@ The example below shows how to use a metric in your `LightningModule <https://py
...
-
-            def validation_step(self, batch, batch_idx, dataloader_idx):
-                x, y = batch
-                preds = self(x)
-                ...
-                self.val_acc[dataloader_idx](preds, y)
-                self.log('val_acc', self.val_acc[dataloader_idx])
-
+Metric logging in Lightning happens through the `self.log` or `self.log_dict` method. Both methods only supports logging of *scalar-tensors*.
+While the vast majority of metrics in torchmetrics returns a scalar tensor, some metrics such as :class:`~torchmetrics.ConfusionMatrix`, :class:`~torchmetrics.ROC`,
```
Comment (Borda) [inline docs/source/pages/lightning.rst:53]: These refs looks outdated

File: docs/source/pages/lightning.rst [hunk 8]
```diff
@@ -118,85 +96,125 @@ If ``on_epoch`` is True, the logger automatically logs the end of epoch metric v
...
 
-                # this will not reset the metric automatically at the epoch end so you
-                # need to call it yourself
-                mean_precision = self.precision.compute()
-                self.log('train_epoch_precision', mean_precision)
-                self.precision.reset()
+* If using metrics in data parallel mode (dp), the metric update/logging should be done
+  in the ``<mode>_step_end`` method (where ``<mode>`` is either ``training``, ``validation``
+  or ``test``). This is due to metric states else being destroyed after each forward pass,
+  leading to wrong accumulation. In practice do the following:
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:164]:
```suggestion
  or ``test``). This is because ``dp`` split the batches during the forward pass and metric states are destroyed after each forward pass, thus leading to wrong accumulation. In practice do the following:
```

File: docs/source/pages/lightning.rst [hunk 9]
```diff
@@ -118,85 +96,125 @@ If ``on_epoch`` is True, the logger automatically logs the end of epoch metric v
...
-    leading to wrong accumulation. In practice do the following:
+        def training_step(self, batch, batch_idx):
+            data, target = batch
+            preds = self(data)
+            # ...
+            return {'loss': loss, 'preds': preds, 'target': target}
 
-    .. testcode:: python
+        def training_step_end(self, outputs):
+            #update and log
```
Comment (rohitgr7) [inline docs/source/pages/lightning.rst:177]:
```suggestion
            # update and log
```

**Review by rohitgr7 — APPROVED** [review]
great improvement!

File: docs/source/pages/lightning.rst [hunk 10]
```diff
@@ -45,45 +49,19 @@ The example below shows how to use a metric in your `LightningModule <https://py
...
-            def validation_step(self, batch, batch_idx, dataloader_idx):
-                x, y = batch
-                preds = self(x)
-                ...
-                self.val_acc[dataloader_idx](preds, y)
-                self.log('val_acc', self.val_acc[dataloader_idx])
-
+Metric logging in Lightning happens through the `self.log` or `self.log_dict` method. Both methods only supports logging of *scalar-tensors*.
+While the vast majority of metrics in torchmetrics returns a scalar tensor, some metrics such as :class:`~torchmetrics.ConfusionMatrix`, :class:`~torchmetrics.ROC`,
+:class:`~torchmetrics.MAP`, :class:`~torchmetrics.RougeScore` return outputs that are non-scalar tensors (often dicts or list of tensors) and should therefore be
```
Comment (SkafteNicki) [inline docs/source/pages/lightning.rst:54]:
```suggestion
While the vast majority of metrics in torchmetrics returns a scalar tensor, some metrics such as :class:`~torchmetrics.ConfusionMatrix`, :class:`~torchmetrics.ROC`,
:class:`~torchmetrics.MeanAveragePrecision`, :class:`~torchmetrics.ROUGEScore` return outputs that are non-scalar tensors (often dicts or list of tensors) and should therefore be
```

File: docs/source/pages/lightning.rst (hunk 7 above)
Comment (SkafteNicki) [reply docs/source/pages/lightning.rst:53]: should be updated now :]

**Review by Borda — APPROVED** [review]
